File: src/game/physics.ts
import { MACHINES } from './machines';
import {
  angle,
  at,
  clamp,
  crossedGate,
  distance,
  finishSide,
  nearest,
  onRoad,
  withinFinish,
} from './geometry';
import type { Input, Point, RaceEvent, Racer, Track } from './types';
export const STEP = 1 / 120;
export const COLORS = ['#078bff', '#ff3a89', '#06a37b', '#ec9b05'] as const;
const NAMES = ['ビリリ', 'さくら', 'みどり', 'ひかる'] as const;

export function spawn(t: Track, r: Racer): void {
  const p = t.path[0]!,
    lateral = [-25, 25, -25, 25][r.id] ?? 0,
    back = r.id >= 2 ? 48 : 0;
  r.x = p.x - Math.sin(p.angle) * lateral - Math.cos(p.angle) * back;
  r.y = p.y + Math.cos(p.angle) * lateral - Math.sin(p.angle) * back;
  if (!onRoad(t, r.x, r.y)) {
    r.x = p.x;
    r.y = p.y;
  }
  r.heading = p.angle;
  r.vx = 0;
  r.vy = 0;
  r.speed = 0;
  r.charge = 0;
  r.boost = 0;
  r.pushing = false;
  r.gate = 0;
  r.progress = 0;
  r.travel = 0;
  r.winding = 0;
}

export function createRacer(
  t: Track,
  id: number,
  color: number,
  cpu: boolean,
): Racer {
  const r: Racer = {
    id,
    color,
    cpu,
    name: cpu ? NAMES[color]! : 'あなた',
    x: 0,
    y: 0,
    heading: 0,
    vx: 0,
    vy: 0,
    speed: 0,
    charge: 0,
    boost: 0,
    pushing: false,
    gate: 0,
    progress: 0,
    hits: 0,
    respawn: 0,
    shock: 0,
    finish: null,
    dashCount: 0,
    travel: 0,
    winding: 0,
  };
  spawn(t, r);
  return r;
}

export function reset(t: Track, r: Racer): void {
  r.shock = 0;
  r.hits++;
  spawn(t, r);
  r.respawn = 1.05;
}

export function advance(t: Track, r: Racer, old: Point): number | null {
  const gate = t.gates[r.gate];
  if (gate && crossedGate(old, r, gate)) r.gate++;
  const n = nearest(t, r),
    maxS = t.gates[r.gate]?.s ?? t.length;
  r.progress = clamp(Math.min(n.s, maxS) / t.length, 0, 1);
  r.travel += distance(old, r);
  // Signed winding rejects collecting sectors by backtracking. Anchors lie in
  // an impassable inner island shared by every supported branch.
  const [ax, ay] = t.lapAnchor;
  r.winding += angle(
    Math.atan2(r.y - ay, r.x - ax) - Math.atan2(old.y - ay, old.x - ax),
  );
  const before = finishSide(t, old),
    after = finishSide(t, r);
  // Every ordered gate, legal travel, and the forward crossing are required together.
  if (
    r.gate !== t.gates.length ||
    r.winding < Math.PI * 2 ||
    r.travel <= t.length * 0.48 ||
    before >= 0 ||
    after < 0 ||
    !withinFinish(t, r)
  )
    return null;
  return -before / (after - before);
}

export function cpuInput(
  t: Track,
  r: Racer,
  difficulty: 'easy' | 'normal',
): Input {
  const n = nearest(t, r),
    target = at(t, n.s + 68),
    err = angle(Math.atan2(target.y - r.y, target.x - r.x) - r.heading);
  const bend = Math.abs(angle(at(t, n.s + 125).angle - n.angle));
  return {
    steer: clamp(err * 2.6, -1, 1),
    push: (Math.abs(err) > 0.36 || bend > 0.55) && r.speed > 70,
    throttle: (difficulty === 'easy' ? 0.5 : 0.67) + 0.035 * r.id,
    assist: false,
  };
}

export function tick(
  t: Track,
  r: Racer,
  input: Input,
  dt: number,
  time: number,
): RaceEvent | null {
  if (r.finish !== null) return null;
  if (r.shock > 0) {
    r.shock = Math.max(0, r.shock - dt);
    if (r.shock === 0) {
      spawn(t, r);
      r.respawn = 0.75;
    }
    return null;
  }
  if (r.respawn > 0) {
    r.respawn = Math.max(0, r.respawn - dt);
    return null;
  }
  const machine = MACHINES[r.color]!;
  const old = { x: r.x, y: r.y },
    steer = clamp(input.steer, -1, 1);
  let event: RaceEvent | null = null;
  if (!input.push && r.pushing) {
    if (r.charge > 0.12) {
      r.boost = (0.35 + 1.15 * r.charge) * machine.duration;
      r.dashCount++;
      event = { type: 'boost', racer: r.id };
    }
    r.charge = 0;
  }
  r.pushing = input.push;
  r.boost = Math.max(0, r.boost - dt);
  if (input.push) {
    r.charge = clamp(r.charge + dt / machine.charge, 0, 1);
    r.boost = 0;
  }
  const target = input.push
    ? 52
    : t.speed *
      machine.speed *
      (input.throttle ?? 1) *
      (r.boost > 0 ? machine.boost : 1);
  r.speed += (target - r.speed) * (1 - Math.exp(-(input.push ? 7 : 1.6) * dt));
  let turn =
    steer *
    machine.turn *
    (input.push ? 3.45 : 2.55) *
    (0.55 + 0.45 * clamp(r.speed / 85, 0, 1));
  if (input.assist) {
    const n = nearest(t, r),
      q = at(t, n.s + 80),
      error = angle(Math.atan2(q.y - r.y, q.x - r.x) - r.heading);
    if (Math.abs(error) < 1.2)
      turn += clamp(error * 1.2, -0.82, 0.82) * (1 - Math.abs(steer) * 0.65);
  }
  r.heading = angle(r.heading + turn * dt);
  const traction =
    1 - Math.exp(-(input.push ? 10 : t.grip * machine.grip) * dt);
  r.vx += (Math.cos(r.heading) * r.speed - r.vx) * traction;
  r.vy += (Math.sin(r.heading) * r.speed - r.vy) * traction;
  r.x += r.vx * dt;
  r.y += r.vy * dt;
  if (!onRoad(t, r.x, r.y)) {
    const hit = { x: r.x, y: r.y };
    r.hits++;
    r.x = old.x;
    r.y = old.y;
    r.speed = 0;
    r.vx = 0;
    r.vy = 0;
    r.charge = 0;
    r.boost = 0;
    r.pushing = false;
    r.gate = 0;
    r.progress = 0;
    r.travel = 0;
    r.winding = 0;
    r.shock = 0.3;
    return { type: 'collision', racer: r.id, ...hit };
  }
  const crossing = advance(t, r, old);
  if (crossing !== null) {
    r.finish = time - dt + crossing * dt;
    r.progress = 1;
    r.speed = 0;
    r.vx = 0;
    r.vy = 0;
    return { type: 'finish', racer: r.id };
  }
  return event;
}

File: backend/src/app/core/physics.py
import json
import math
from pathlib import Path

from pydantic import BaseModel

from .models import Input, Racer

STEP = 1 / 120


class Point(BaseModel):
    x: float
    y: float
    s: float = 0
    angle: float = 0
    width: float = 115


class Finish(BaseModel):
    a: tuple[float, float]
    b: tuple[float, float]
    normal: tuple[float, float]


class Track(BaseModel):
    id: int
    name: str
    speed: float
    grip: float
    wallRise: float
    lapAnchor: tuple[float, float]
    length: float
    path: list[Point]
    gates: list[Point]
    outer: list[tuple[float, float]]
    holes: list[list[tuple[float, float]]]
    finish: Finish


class Machine(BaseModel):
    speed: float
    duration: float
    charge: float
    boost: float
    turn: float
    grip: float


class Data(BaseModel):
    tracks: list[Track]
    machines: list[Machine]


data = Data.model_validate(
    json.loads((Path(__file__).parents[1] / "data/game.gen.json").read_text())
)
TRACKS, MACHINES = data.tracks, data.machines


def clamp(v: float, a: float, b: float) -> float:
    return max(a, min(b, v))


def angle(v: float) -> float:
    return math.atan2(math.sin(v), math.cos(v))


def on_road(t: Track, x: float, y: float) -> bool:
    def inside(poly: list[tuple[float, float]]) -> bool:
        hit = False
        for i, a in enumerate(poly):
            b = poly[i - 1]
            if (a[1] > y) != (b[1] > y) and x < (b[0] - a[0]) * (y - a[1]) / (b[1] - a[1]) + a[0]:
                hit = not hit
        return hit

    def clearance(poly: list[tuple[float, float]], rise: float = 0) -> float:
        best = math.inf
        for i, a in enumerate(poly):
            b = poly[(i + 1) % len(poly)]
            dx, dy = b[0] - a[0], b[1] - a[1]
            u = clamp(((x - a[0]) * dx + (y + rise - a[1]) * dy) / (dx * dx + dy * dy or 1), 0, 1)
            best = min(best, math.hypot(x - a[0] - u * dx, y + rise - a[1] - u * dy))
        return best

    return (
        inside(t.outer)
        and clearance(t.outer) >= 30
        and all(not inside(h) and clearance(h, t.wallRise) >= 48 for h in t.holes)
    )


def spawn(t: Track, r: Racer) -> None:
    p = t.path[0]
    back = (r.id // 2) * 24
    lanes = [
        v
        for v in (-25, 25, 0, -50, 50)
        if on_road(
            t,
            p.x - math.sin(p.angle) * v - math.cos(p.angle) * back,
            p.y + math.cos(p.angle) * v - math.sin(p.angle) * back,
        )
    ]
    lateral = lanes[r.id % 2] if len(lanes) > r.id % 2 else (lanes[0] if lanes else 0)
    r.x = p.x - math.sin(p.angle) * lateral - math.cos(p.angle) * back
    r.y = p.y + math.cos(p.angle) * lateral - math.sin(p.angle) * back
    if not on_road(t, r.x, r.y):
        r.x, r.y = p.x, p.y
    r.heading = p.angle
    r.vx = r.vy = r.speed = r.charge = r.boost = r.progress = r.travel = r.winding = 0
    r.pushing = False
    r.gate = 0


def create(t: Track, slot: int, color: int, name: str) -> Racer:
    r = Racer(id=slot, color=color, name=name)
    spawn(t, r)
    return r


def nearest(t: Track, r: Racer) -> Point:
    return min(
        (
            p
            for p in t.path
            if not (r.gate == 0 and p.s > t.length * 0.85)
            and not (r.gate == len(t.gates) and p.s < t.length * 0.15)
        ),
        key=lambda p: (p.x - r.x) ** 2 + (p.y - r.y) ** 2,
    )


def at(t: Track, s: float) -> Point:
    low, high = 0, len(t.path) - 1
    while low < high:
        m = (low + high) // 2
        if t.path[m].s < s:
            low = m + 1
        else:
            high = m
    return t.path[low]


def tick(t: Track, r: Racer, inp: Input, time: float) -> None:
    dt = STEP
    if r.finish is not None:
        return
    if r.shock > 0:
        r.shock = max(0, r.shock - dt)
        if r.shock == 0:
            spawn(t, r)
            r.respawn = 0.75
        return
    if r.respawn > 0:
        r.respawn = max(0, r.respawn - dt)
        return
    m = MACHINES[r.color]
    ox, oy = r.x, r.y
    if not inp.push and r.pushing:
        if r.charge > 0.12:
            r.boost = (0.35 + 1.15 * r.charge) * m.duration
            r.dashCount += 1
        r.charge = 0
    r.pushing = inp.push
    r.boost = max(0, r.boost - dt)
    if inp.push:
        r.charge = clamp(r.charge + dt / m.charge, 0, 1)
        r.boost = 0
    target = 52 if inp.push else t.speed * m.speed * (m.boost if r.boost > 0 else 1)
    r.speed += (target - r.speed) * (1 - math.exp(-(7 if inp.push else 1.6) * dt))
    turn = (
        inp.steer
        * m.turn
        * (3.45 if inp.push else 2.55)
        * (0.55 + 0.45 * clamp(r.speed / 85, 0, 1))
    )
    if inp.assist:
        q = at(t, nearest(t, r).s + 80)
        error = angle(math.atan2(q.y - r.y, q.x - r.x) - r.heading)
        if abs(error) < 1.2:
            turn += clamp(error * 1.2, -0.82, 0.82) * (1 - abs(inp.steer) * 0.65)
    r.heading = angle(r.heading + turn * dt)
    traction = 1 - math.exp(-(10 if inp.push else t.grip * m.grip) * dt)
    r.vx += (math.cos(r.heading) * r.speed - r.vx) * traction
    r.vy += (math.sin(r.heading) * r.speed - r.vy) * traction
    r.x += r.vx * dt
    r.y += r.vy * dt
    if not on_road(t, r.x, r.y):
        r.hits += 1
        r.x, r.y = ox, oy
        r.speed = r.vx = r.vy = r.charge = r.boost = r.progress = r.travel = r.winding = 0
        r.pushing = False
        r.gate = 0
        r.shock = 0.3
        return
    if r.gate < len(t.gates):
        g = t.gates[r.gate]
        tx, ty = math.cos(g.angle), math.sin(g.angle)
        if (
            (ox - g.x) * tx + (oy - g.y) * ty < 0
            and (r.x - g.x) * tx + (r.y - g.y) * ty >= 0
            and math.hypot(r.x - g.x, r.y - g.y) < g.width
        ):
            r.gate += 1
    n = nearest(t, r)
    limit = t.gates[r.gate].s if r.gate < len(t.gates) else t.length
    r.progress = clamp(min(n.s, limit) / t.length, 0, 1)
    r.travel += math.hypot(r.x - ox, r.y - oy)
    ax, ay = t.lapAnchor
    r.winding += angle(math.atan2(r.y - ay, r.x - ax) - math.atan2(oy - ay, ox - ax))
    f = t.finish
    mx, my = (f.a[0] + f.b[0]) / 2, (f.a[1] + f.b[1]) / 2
    before = (ox - mx) * f.normal[0] + (oy - my) * f.normal[1]
    after = (r.x - mx) * f.normal[0] + (r.y - my) * f.normal[1]
    dx, dy = f.b[0] - f.a[0], f.b[1] - f.a[1]
    dot = (r.x - f.a[0]) * dx + (r.y - f.a[1]) * dy
    if (
        r.gate == len(t.gates)
        and r.winding >= math.pi * 2
        and r.travel > t.length * 0.48
        and before < 0 <= after
        and 0 <= dot <= dx * dx + dy * dy
    ):
        r.finish = time - dt - before / (after - before) * dt
        r.progress = 1
        r.speed = r.vx = r.vy = 0
